#meta_fams_inds_vcf_5pc_diff_dens.R

#inverse variance-weighted meta analysis combining the family TDT results with the indepedent individual results

extension="final_collection"

#Readin family results as generated by Cassie:
cassiedir<-"/nv/vol185/MEGA/release4/IMPUTED_TOPMED/family_analysis/"
cohorts1=c("EAS", "AFR","EUR","AMR", "FIN")
cohorts=c("AFR","EUR","AMR", "FIN")

readinfam<-function(cohort, chromosome){
r<-read.table(paste0(cassiedir,"/",cohort,"/chr",chromosome,".filter_maf_gt_0.005_rsq_gt_0.8.relateds_with_pheno.tdt"), header=T)
return(r)
}
getallfam<-function(cohort){
g<-lapply(c(1:22),cohort=cohort, readinfam)
g<-do.call("rbind",g)
return(g)
}
results<-lapply(cohorts1,getallfam)
names(results)<-cohorts1



#Read in independent results as generated by Jamie:
jamiedir<-paste0("/scratch/ji3x/",extension,"_gwas/META/META/")
ind<-get(load(paste0(jamiedir,"resultsmeta_5pcs_vcf_5pc_diff_dens.RData")))
rownames(ind)<-ind$snp.name
allsnps<-rownames(ind)


ind$SNP<-ind$snp.name
ind$RefAllele<-ind$Allele1
ind$NonRefAllele<-ind$Allele2
ind$BETA<-ind$Effect
ind$SE<-ind$StdErr

getbetas<-function(cohort){
fam<-results[[cohort]]
fam[,paste0("BETA_",cohort)]<-log(fam$OR)
fam[,paste0("SE_",cohort)]<-sqrt((1/fam$T)+(1/fam$U))
fam[,paste0("RefAllele_",cohort)]<-fam$A2
fam[,paste0("NonRefAllele_",cohort)]<-fam$A1
#remove dodgy SNPs (mendel rate 1% for families):
fam[,paste0("BETA_",cohort)]<-ifelse(is.infinite(fam[,paste0("BETA_",cohort)]),NA,fam[,paste0("BETA_",cohort)])
fam[,paste0("SE_",cohort)]<-ifelse(is.infinite(fam[,paste0("SE_",cohort)]),NA,fam[,paste0("SE_",cohort)])
men<-read.table(file=paste0(cassiedir,"/",cohort,"/snps_with_mendelInconsistencies_gt1pct.txt"), header=T, as.is=T)
fam<-fam[!fam$SNP %in% men$SNP,]
try<-c("BETA_","SE_","RefAllele_","NonRefAllele_")
try<-paste0(try,cohort)
try<-c("SNP",try)
fam<-fam[,colnames(fam) %in% try]
return(fam)
}
fam<-lapply(cohorts1, getbetas)
names(fam)<-cohorts1

indi<-merge(ind, fam[[cohorts1[1]]], by="SNP", all.x=T)
indi<-merge(indi, fam[[cohorts1[2]]], by="SNP", all.x=T)
indi<-merge(indi, fam[[cohorts1[3]]], by="SNP", all.x=T)
indi<-merge(indi, fam[[cohorts1[4]]], by="SNP", all.x=T)
indi<-merge(indi, fam[[cohorts1[5]]], by="SNP", all.x=T)

ind<-ind[,c("SNP","RefAllele","NonRefAllele",
"BETA", "SE")]

writeout<-function(cohort){
try<-c("RefAllele_","NonRefAllele_","BETA_","SE_")
try<-paste0(try,cohort)
fam1<-indi[,c("SNP",try)]
rownames(fam1)<-fam1$SNP
colnames(fam1)<-c("SNP","RefAllele","NonRefAllele",
"BETA", "SE")
write.table(fam1, file=paste0(jamiedir,"meta_file_fam_",cohort,"_newqc_1_dens.tbl"), 
sep=" ", col.names=T, row.names=F, quote=F)
}

write.table(ind, file=paste0(jamiedir, "meta_file_ind_vcf_5pc_diff_dens.tbl"), sep=" ", col.names=T, row.names=F, quote=F)
lapply(cohorts1, writeout)

#now generate a file to use in METAL:
generatefile<-function(extension){
sink(file=paste0("~/programs/",extension,"/META/metal_script_fam_ind_vcf_5pc_diff_dens"))
cat(paste0("#metal_script_fam_ind_vcf_5pc_diff_dens

#THIS SCRIPT EXECUTES AN ANALYSIS OF THE FAMILY AND CASE-CONTROL GROUPS

#LOAD THE INPUT FILES

# === DESCRIBE AND PROCESS THE FIRST INPUT FILE ===
MARKER SNP
ALLELE RefAllele NonRefAllele
EFFECT BETA
STDERR SE
SCHEME STDERR
PROCESS ",jamiedir,"/meta_file_ind_vcf_5pc_diff_dens.tbl

#All files are of the same format so can process immediately:
PROCESS ",jamiedir,"/meta_file_fam_AFR_newqc_1_dens.tbl
PROCESS ",jamiedir,"/meta_file_fam_EUR_newqc_1_dens.tbl
PROCESS ",jamiedir,"/meta_file_fam_AMR_newqc_1_dens.tbl
PROCESS ",jamiedir,"/meta_file_fam_FIN_newqc_1_dens.tbl
PROCESS ",jamiedir,"/meta_file_fam_EAS_newqc_1_dens.tbl

OUTFILE \'",jamiedir,"/METAANALYSIS_ind_fam_tdt_vcf_5pc_diff_dens\'
ANALYZE

QUIT"))
sink()
}
generatefile("final_collection")


#finally the file to run the metal script:

generatescript<-function(extension){
sink(file=paste0("~/programs/",extension,"/META/run_metal_ind_fam_vcf_5pc_diff_dens.sh"))
cat(paste0("#!/bin/bash
#run_metal_ind_fam_vcf_5pc_diff_dens

/home/ji3x/software/generic-metal/metal metal_script_fam_ind_vcf_5pc_diff_dens

#Currently can't get the results to output to where i'd like them to,
#so I am moving here:
mv METAANALYSIS1.TBL METAANALYSIS_ind_fam_tdt_vcf_5pc_diff_dens.TBL
mv METAANALYSIS1.TBL.info METAANALYSIS_ind_fam_tdt_vcf_5pc_diff_dens.TBL.info

mv METAANALYSIS_ind_fam_tdt_vcf_5pc_diff_dens.TBL ",jamiedir,"
mv METAANALYSIS_ind_fam_tdt_vcf_5pc_diff_dens.TBL.info ",jamiedir,"
"))
sink()
system(paste0("chmod a=rwx ~/programs/",extension,"/META/run_metal_ind_fam_vcf_5pc_diff_dens.sh"))
}
generatescript("final_collection")

